### PR TITLE
As/cached router implementation

### PR DESCRIPTION
**Pull request by anbitasiregar**
To test:
- Create a network (ie. open terminal and input the following command: dns-sd -R ThreadNet_ios _thread-net._udp local. 19779)
- Run the app and connect to the network
- The app should log the new cached router after successful connection.

### CODE REVIEW

File: ThreadGroup/Controllers/TGMainViewController.m [hunk 1]
```diff
@@ -124,19 +127,47 @@ - (void)commonInit {
...
+    [[TGKeychainManager sharedManager] saveRouterItem:item withCompletion:^(NSError *error) {
+        if (error) {
+            NSLog(@"Error saving router to keychain: %@", error);
+        }
+        NSLog(@"New cached router: %@", self.cachedRouter.name);
+    }];
+}
+
+- (TGRouter *)cachedRouter {
+    return [[TGKeychainManager sharedManager] getRouterItem];
```
Comment (nservidio) [inline ThreadGroup/Controllers/TGMainViewController.m:167]: Do you need both "cachedRouter" and "getCachedRouter"?

File: ThreadGroup/Controllers/TGRouterAuthViewController.h [hunk 2]
```diff
@@ -12,11 +12,13 @@
 @class TGRouterAuthViewController;
 
 @protocol TGRouterAuthViewControllerDelegate <NSObject>
-- (void)routerAuthenticationSuccessful:(TGRouterAuthViewController *)routerAuthenticationView;
 - (void)routerAuthenticationCanceled:(TGRouterAuthViewController *)routerAuthenticationView;
+- (void)okButtonWasPressedInRouterAuthentication:(TGRouterAuthViewController *)routerAuthenticationView;
```
Comment (nservidio) [inline ThreadGroup/Controllers/TGRouterAuthViewController.h]: I would standardize these delegate method names, i.e

routerAuthenticationViewControllerDidPressOkButton:(TGRouterAuthViewController *)routerAuthenticationView.

I would even update the existing methods, but maybe check with Pat or Lu first.

File: ThreadGroup/Controllers/TGRouterAuthViewController.h [hunk 3]
```diff
@@ -12,11 +12,13 @@
...
 @end
 
 @interface TGRouterAuthViewController : UIViewController
 @property (nonatomic, weak) id<TGRouterAuthViewControllerDelegate> delegate;
 @property (nonatomic, strong) TGRouter *item;
+
+- (void)authenticationFailedState;
```
Comment (nservidio) [inline ThreadGroup/Controllers/TGRouterAuthViewController.h]: I think this method could have a slightly more descriptive name. It sounds like a BOOL value that tells you if you failed authentication. Something like: - (void)updateUIForFailedAuthentication;

File: ThreadGroup/Controllers/TGRouterAuthViewController.h (hunk 2 above)
Comment (luquantan) [inline ThreadGroup/Controllers/TGRouterAuthViewController.h]: I agree. :+1:

File: ThreadGroup/Controllers/TGRouterAuthViewController.h (hunk 3 above)
Comment (luquantan) [inline ThreadGroup/Controllers/TGRouterAuthViewController.h]: To this as well. :+1:

File: ThreadGroup/Controllers/TGMainViewController.m (hunk 1 above)
Comment (anbitasiregar) [inline ThreadGroup/Controllers/TGMainViewController.m:167]: The cachedRouter getter replaced the getCachedRouter method.

**Pull request merged**